File: language/runtime/apple/core/Sources/RuntimeHostAppleCore/Module/Media/MediaRequests.swift
import Foundation

/// The Apple media request surface for one runtime host embedder.
@MainActor
public protocol MediaRequests: AnyObject {
  /// List one page of media assets for one query.
  func listMedia(
    _ request: RuntimeHostMediaListRequest
  ) -> RuntimeHostMediaListResponse

  /// Read one media asset descriptor by stable identifier.
  func readMedia(
    identifier: String
  ) -> RuntimeHostMediaReadResponse

  /// Import one local path into the media library.
  func importMediaPath(
    _ request: RuntimeHostMediaImportPathRequest
  ) -> RuntimeHostMediaImportPathResponse

  /// Delete one batch of media assets and return the deleted count.
  func deleteMedia(
    _ request: RuntimeHostMediaDeleteRequest
  ) -> RuntimeHostMediaDeleteResponse
}

File: language/runtime/android/kotlin/src/main/kotlin/dev/destack/runtime/android/module/media/MediaRequests.kt
package dev.destack.runtime.android.module.media

import dev.destack.runtime.android.core.hostStatusNotSupported

/**
 * The media request surface attached to one Android runtime host.
 */
public interface MediaRequests {
    /**
     * List one page of media assets for one query.
     */
    public fun listMedia(
        request: RuntimeHostMediaListRequest,
    ): RuntimeHostMediaListResponse

    /**
     * Read one media asset descriptor by stable identifier.
     */
    public fun readMedia(
        identifier: String,
    ): RuntimeHostMediaReadResponse

    /**
     * Import one local path into the media library.
     */
    public fun importMediaPath(
        request: RuntimeHostMediaImportPathRequest,
    ): RuntimeHostMediaImportPathResponse

    /**
     * Delete one batch of media assets and return the deleted count.
     */
    public fun deleteMedia(
        request: RuntimeHostMediaDeleteRequest,
    ): RuntimeHostMediaDeleteResponse
}

/**
 * The explicit unsupported media request surface for one Android runtime host.
 */
public object UnsupportedMediaRequests : MediaRequests {
    override fun listMedia(
        request: RuntimeHostMediaListRequest,
    ): RuntimeHostMediaListResponse {
        return RuntimeHostMediaListResponse(status = hostStatusNotSupported)
    }

    override fun readMedia(
        identifier: String,
    ): RuntimeHostMediaReadResponse {
        return RuntimeHostMediaReadResponse(status = hostStatusNotSupported)
    }

    override fun importMediaPath(
        request: RuntimeHostMediaImportPathRequest,
    ): RuntimeHostMediaImportPathResponse {
        return RuntimeHostMediaImportPathResponse(status = hostStatusNotSupported)
    }

    override fun deleteMedia(
        request: RuntimeHostMediaDeleteRequest,
    ): RuntimeHostMediaDeleteResponse {
        return RuntimeHostMediaDeleteResponse(status = hostStatusNotSupported)
    }
}
